### DISCUSSION TITLE
How to get exception details in browser?

### DISCUSSION THREAD

**Discussion by elringus** [unoplatform/Uno.Wasm.Bootstrap#437] [Q&A] [answered]
I'm getting the following message in the browser console:

```
Debugger.Debug: handle exception - 0 - 0x1046938 - 0x5a4838 - 0x5a47c0 (dotnet.js:1:16820)
```

Is there a way to find out the type of the exception and stacktrace? The project is built under debug configuration and I've tried adding the following to the .csproj:

```
<PropertyGroup Condition="'$(Configuration)'=='Debug'">
    <MonoRuntimeDebuggerEnabled>true</MonoRuntimeDebuggerEnabled>
    <DebugType>portable</DebugType>
    <DebugSymbols>true</DebugSymbols>
    <DefineConstants>$(DefineConstants);TRACE;DEBUG</DefineConstants>
</PropertyGroup>
```

— but it didn't help.

**Comment by jeromelaban** [comment, marked as answer]
Those are generally handled exceptions, but you can still tentatively get those [to show up using this](https://github.com/unoplatform/Uno.Wasm.Bootstrap#aot-debugging-and-mono-tracing-net-5-only).

If you're using AOT, you likely won't get a stack trace on the managed side, but you may get the Wasm stack trace when expanding the error message in the browser logs (if the message happened while the debugger was opened).

**Reply by elringus** [reply]
I've tried explicitly setting interpreter mode, so my properties are now as follows:

```
<PropertyGroup Condition="'$(Configuration)'=='Debug'">
    <WasmShellMonoRuntimeExecutionMode>Interpreter</WasmShellMonoRuntimeExecutionMode>
    <MonoRuntimeDebuggerEnabled>true</MonoRuntimeDebuggerEnabled>
    <DebugType>portable</DebugType>
    <DebugSymbols>true</DebugSymbols>
    <DefineConstants>$(DefineConstants);TRACE;DEBUG</DefineConstants>
</PropertyGroup>
```

— but it still printed the same debug message.

I'm pretty sure those are not handled exceptions, as my code is not running as expected. I'm receiving those debug messages instead of the console.prints assigned to websocket message callback:

```csharp
private static void Main (string[] args)
{
    Task.Run(ListenServerAsync);
}

private static async Task ListenServerAsync ()
{
    var client = new Client();
    var status = await client.ConnectAsync(41016);
    client.Subscribe<IServerMessage>(Console.WriteLine);
    await status.MaintainTask;
}
```

**Reply by jeromelaban** [reply]
Have you tried adding the code specified in the documentation:

```csharp
MonoInternals.mono_trace_enable(1);
MonoInternals.mono_trace_set_options("E:all");
```

Handled exceptions generally don't show up any other way.

**Reply by elringus** [reply]
I thought they're for AOT mode only. I've added those and the MonoInternals class, but now the following is logged in the browser console on startup:

```
Error: System.DllNotFoundException: __Native
   at Program.Main(String[] args) in C:\...\Program.cs:line 21
    call_method http://localhost:5000/package_4a3d70c8acbaef988024fad526b820f9fe63f354/uno-bootstrap.js:157
    mainInit http://localhost:5000/package_4a3d70c8acbaef988024fad526b820f9fe63f354/uno-bootstrap.js:198
    initializeRequire http://localhost:5000/package_4a3d70c8acbaef988024fad526b820f9fe63f354/uno-bootstrap.js:479
    init http://localhost:5000/package_4a3d70c8acbaef988024fad526b820f9fe63f354/uno-bootstrap.js:172
    onRuntimeInitialized http://localhost:5000/package_4a3d70c8acbaef988024fad526b820f9fe63f354/uno-bootstrap.js:38
    _finalize_startup http://localhost:5000/package_4a3d70c8acbaef988024fad526b820f9fe63f354/dotnet.js:1
    onPendingRequestComplete http://localhost:5000/package_4a3d70c8acbaef988024fad526b820f9fe63f354/dotnet.js:1
    processFetchResponseBuffer http://localhost:5000/package_4a3d70c8acbaef988024fad526b820f9fe63f354/dotnet.js:1
uno-bootstrap.js:201:21
```

The 21th line is `MonoInternals.mono_trace_enable(1);`

**Reply by jeromelaban** [reply]
The documentation is lacking a bit when not running with AOT, but you can enable native linking with the following:

```xml
<ItemGroup>
    <WasmShellAdditionalPInvokeLibrary Include="libc" />
</ItemGroup>
```

The value in the include is not important, but this one will do to enable native linking.

**Reply by elringus** [reply]
Thank you for the info! Unfortunately, it's now throwing the following on build (even though I have WSL with Ubuntu 18.04 installed):

```
Uno.Wasm.Bootstrap.targets(173, 5): System.InvalidCastException: The Windows Subsystem for Linux is not installed, please install Ubuntu 18.04 by visiting https://docs.microsoft.com/en-us/windows/wsl/install-win10.
 at Uno.Wasm.Bootstrap.ShellTask_vdcaa0c68ff43a79006ac22ad1707f768e02abc4d.ValidateEmscripten() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 1009
 at Uno.Wasm.Bootstrap.ShellTask_vdcaa0c68ff43a79006ac22ad1707f768e02abc4d.RunPackager() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 719
 at Uno.Wasm.Bootstrap.ShellTask_vdcaa0c68ff43a79006ac22ad1707f768e02abc4d.Execute() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 225
```

**Reply by jeromelaban** [reply]
Make sure to update to the latest boostrapper 3.0.0, which does not require WSL.

**Reply by elringus** [reply]
Thank you, I was able to get the exception info:

```
[0x1:] EXCEPTION handling: System.InvalidOperationException: XmlConstructorInaccessible
```

You were right that it's handled, though it's behaving not as expected. It looks like `System.Xml.Serialization` fails to deserialize a type specified in an assembly used by the main program, eg:

```csharp
using (var stringReader = new StringReader(xml))
    return new XmlSerializer(dataType).Deserialize(stringReader);
```

When I invoke this code in the main program, it runs ok, but when the same thing is invoked from an assembly, it throws `XmlConstructorInaccessible` exception.

**Reply by jeromelaban** [reply]
Good. This new error generally means that the linker is removing important pieces of the deserialized type. Try disabling the linker or add exclusions for the type being deserialized (in the LinkerConfig.xml file).

**Reply by elringus** [reply]
Got it working. Thank you very much for all the help!
